### ISSUE TITLE
Vector subscript out of range on "Vector<Number>::add(..)".

### ISSUE THREAD

**Issue by tulioligneul** [dealii/dealii#4916]
In the methods:
```
Vector<Number>::add (const std::vector<size_type> &indices, const Vector<OtherNumber>  &values)
Vector<Number>::add (const std::vector<size_type> &indices, const std::vector<OtherNumber>  &values)
```
A pointer to the first element of the std::vectors is gotten, for instance, by &indices[0], which implies that the first element of the vector exists. If an empty vector is used, while this works on gcc, this results in a crash when using visual studio. In this case, the pointer should be gotten by [std::vector::data() ](http://en.cppreference.com/w/cpp/container/vector/data).

I'attaching a fix for this. It also includes an update to a CMake script to activate DEAL_II_WITH_CXX11 on Windows.

[myUpdate.txt](https://github.com/dealii/dealii/files/1237296/myUpdate.txt)

**Comment by davydden** [comment]
Would you be willing to open a pull request here with the fix?

**Comment by bangerth** [comment]
Ah, good catch, yes! I think the easier alternative would be to just put an
```
  if (indices.size() > 0)
```
in front of the whole call of the other `add()` function -- there is no point in calling it if there is not going to be anything to do. Do you want to try that?

(In a perfect world, you could even submit a patch for this through github. If that's the first time you do this, take a look at video lecture 32.8 here: http://www.math.colostate.edu/~bangerth/videos.html . It walks you through the steps you have to take.)

On a separate note, your patch also contains changes to cmake files. These are logically separate. Can you describe what they do, and submit those via a separate pull request?

Thanks in advance, and thanks for figuring out the patch you have here!

**Comment by tulioligneul** [comment]
I couldn`t test the size of the vector because the add function was called from inside deal.II. While tweaking my implementation, at some point, it was calling:

```
ConstraintMatrix::
distribute_local_to_global (const FullMatrix<typename MatrixType::value_type>     &local_matrix,
                            const Vector<typename VectorType::value_type>         &local_vector,
                            const std::vector<size_type> &local_dof_indices,
                            MatrixType                   &global_matrix,
                            VectorType                   &global_vector,
                            bool                          use_inhomogeneities_for_rhs) const
```
where the local_vector only had zero entries.  Then, the add function was called with an empty vector.

**Comment by bangerth** [comment]
Right. I was thinking of changing the place where you make your modification from
```
  add (indices.size(), &indices[0], values.val);
```
to
```
  if (indices.size() > 0)
    add (indices.size(), &indices[0], values.val);
```
This should work as well. But regardless, I do like the use of `.data()`. It is clearer than taking the address of the first element.

**Comment by bangerth** [comment]
This is now fixed by #4918 and the follow-ups listed there.
